### PR TITLE
Wrote registry functions

### PR DESCRIPTION
**Pull request by JasperSlimmens**
I wrote the functions for the Registry contract. Have done small tests, but did not create proper test yet. Will do it later. Have a look and tell me what you think!

### CODE REVIEW

**Comment by JasperSlimmens** [comment]
Updated contract and finished test. Test contains a lot of console logging. I like visual confirmation, these can all be commented out.

For the contract I think main functionalities are working. What might be a nice improvement is making distinction between removing and deactivating. This would make it easier to get historical data.

File: contracts/BookkeeperRegistry.sol [hunk 1]
```diff
@@ -13,319 +13,447 @@ import "./Controllable.sol";
...
-    mapping (address => bool) public vaultCheck;
-    mapping (address => bool) public strategyCheck;
-    mapping (address => bool) public rewardPoolCheck;
-    mapping (address => bool) public coreContractCheck;
+    // Keeping track of links between all added contracts.
+    mapping (address => bool) public isActive;
+    mapping (address => uint256) public addedOnBlock;
+    mapping (address => string) public contractType;
+    mapping (address => address) public vaultStrategy;
+    mapping (address => address[]) public vaultStrategies;
```
Comment (belbix) [inline contracts/BookkeeperRegistry.sol:27]: do you need both?

File: contracts/BookkeeperRegistry.sol [hunk 2]
```diff
@@ -13,319 +13,447 @@ import "./Controllable.sol";
...
-    //Change Reward Pool for existing vault. Again vault is needed as input, as no on-chain link.
-    function changeRewardPool(address _rewardPool, address _vault) external onlyGovernance validVault(_vault) {
-      require(_vault != address(0), "vault shouldn't be empty");
-      require(!rewardPoolCheck[_rewardPool], "reward pool already exists");
-      require(_rewardPool != address(0), "new reward pool shouldn't be empty");
-
-      uint i = 0;
-      while (vaultList[i] != _vault) {
+    //Add a vault. RewardPool can be address(0).
+    function addVault(address _vault, address _rewardPool, bool _multipleStrategies) external onlyGovernance {
```
Comment (belbix) [inline contracts/BookkeeperRegistry.sol:89]: let's avoid this solution - _multipleStrategies
we can add only one strategy to the array

File: contracts/BookkeeperRegistry.sol [hunk 3]
```diff
@@ -13,319 +13,447 @@ import "./Controllable.sol";
...
     using SafeMath for uint256;
 
-    // Keeping track of all added contracts. Lists are made so they can be returned by functions.
-    mapping (address => bool) public vaultCheck;
-    mapping (address => bool) public strategyCheck;
-    mapping (address => bool) public rewardPoolCheck;
-    mapping (address => bool) public coreContractCheck;
+    // Keeping track of links between all added contracts.
+    mapping (address => bool) public isActive;
+    mapping (address => uint256) public addedOnBlock;
```
Comment (belbix) [inline contracts/BookkeeperRegistry.sol:24]: it is wrong
we should know when a contract created

File: contracts/BookkeeperRegistry.sol (hunk 2 above)
Comment (belbix) [reply contracts/BookkeeperRegistry.sol:89]: and I think better to have a split function for adding vault, pool and link between them
also, we should have a bulk method with arrays to reduce gas cost (I don't want to pay for ~200+ transactions)

File: contracts/BookkeeperRegistry.sol [hunk 4]
```diff
@@ -13,319 +13,447 @@ import "./Controllable.sol";
...
+        require(keccak256(abi.encodePacked(contractType[_vault])) != keccak256(abi.encodePacked("vault")), "vault already exists");
+
+        address vault = _vault;
+        address strategy = IVault(_vault).strategy();
+        address rewardPool = _rewardPool;
+        address underlying = IVault(_vault).underlying();
+
+        isActive[vault] = true;
+        addedOnBlock[vault] = block.number;
+        contractType[vault] = "vault";
```
Comment (belbix) [inline contracts/BookkeeperRegistry.sol:100]: use int codes with constants

File: contracts/BookkeeperRegistry.sol [hunk 5]
```diff
@@ -13,319 +13,447 @@ import "./Controllable.sol";
...
+        vaultUnderlying[vault] = underlying;
+        vaultMultipleStrategies[vault] = _multipleStrategies;
+
+        if (vaultMultipleStrategies[vault]) {
+          vaultStrategies[vault].push(strategy);
+        } else {
+          vaultStrategy[vault] = strategy;
+        }
+
+        isActive[strategy] = true;
```
Comment (belbix) [inline contracts/BookkeeperRegistry.sol:110]: need to investigate which contracts we should check
maybe we can check only vaults and other entities just unlink

File: contracts/BookkeeperRegistry.sol [hunk 6]
```diff
@@ -13,319 +13,447 @@ import "./Controllable.sol";
...
+    //Add a vault. RewardPool can be address(0).
+    function addVault(address _vault, address _rewardPool, bool _multipleStrategies) external onlyGovernance {
+        require(_vault != address(0), "new vault should not be empty");
+        require(keccak256(abi.encodePacked(contractType[_vault])) != keccak256(abi.encodePacked("vault")), "vault already exists");
+
+        address vault = _vault;
+        address strategy = IVault(_vault).strategy();
+        address rewardPool = _rewardPool;
+        address underlying = IVault(_vault).underlying();
+
```
Comment (belbix) [inline contracts/BookkeeperRegistry.sol:97]: what happens if it is a wrong address?

**Comment by belbix** [comment]
okay, I merge it as a draft version
@rernst76 will continue work with it
please, don't create so huge PRs

File: contracts/BookkeeperRegistry.sol (hunk 1 above)
Comment (JasperSlimmens) [reply contracts/BookkeeperRegistry.sol:27]: I was trying to generalize for the case vaults might have multiple strategies.

**Pull request merged**